src/mars.c

```diff
@@ -1694,21 +1710,74 @@ Language changes listed by -transition=id:\n\
                 fprintf(global.stdmsg, "code      %s\n", m->toChars());
 
             obj_start(m->srcfile->toChars());
-            genObjFile(m, global.params.multiobj);
+            genObjFile(m, false);
             if (entrypoint && m == rootHasMain)
-                genObjFile(entrypoint, global.params.multiobj);
+                genObjFile(entrypoint, false);
             for (size_t j = 0; j < Module::amodules.dim; j++)
             {
+                // todo: This part is not the best for the total size of object files.
+                // For example:
+                // When both of main.d and code.d import mx, even if main.obj
+                // already contains mx->marray, it is stored in code.obj again.
                 Module *mx = Module::amodules[j];
-                if (mx != m && mx->importedFrom == m && (mx->marray || mx->massert || mx->munittest))
-                    genhelpers(mx, true);
+                if (mx == m || mx->importedFrom != m)
+                    continue;
+                if (!mx->marray && !mx->massert && !mx->munittest)
+                    continue;
+                genhelpers(mx, true);
             }
             obj_end(library, m->objfile);
+
+            if (global.errors && !global.params.lib)
+                m->deleteObjFile();
+        }
+    }
+    else
+    {
+        /* global.params.multiobj == true:
+         *  Each compiled symbols are stored in separated doppelganger modules.
+         *  The generated object files have numbered names.
+         *  e.g.
+         *      dmd -lib main.d code.d
+         *      dmd -multiobj main.d code.d
+         *      // main.foo() --> main_1_<hash>.obj
+         *      // main.bar() --> main_2_<hash>.obj
+         *      // code.baz() --> code_3_<hash>.obj
+         *      // std.stdio.writeln!()() --> stdio_4_<hash>.obj
+         *      // std.stdio.__array() --> stdio_5_<hash>.obj   (Bugzilla 14828)
+         *      // --> with -lib, all *.obj will be stored in one .lib file.
+         */
+        for (size_t i = 0; i < modules.dim; i++)
+        {
+            Module *m = modules[i];
+            if (global.params.verbose)
+                fprintf(global.stdmsg, "code      %s\n", m->toChars());
+
+            obj_start(m->srcfile->toChars());
+            genObjFile(m, true);
+            if (entrypoint && m == rootHasMain)
+                genObjFile(entrypoint, true);
+            obj_end(library, m->objfile);
+
             obj_write_deferred(library);
 
             if (global.errors && !global.params.lib)
                 m->deleteObjFile();
         }
+
+        for (size_t j = 0; j < Module::amodules.dim; j++)
+        {
+            Module *mx = Module::amodules[j];
+            if (mx->isRoot())
+                continue;
+            if (!mx->marray && !mx->massert && !mx->munittest)
+                continue;
+
+            genHelpersObjFile(library, mx);
```

Comment:
You could just move the code here, then it's on the same abstraction level as the rest.
```cpp
    const char *mname = m->srcfile->toChars();
    // emit marray/masser/munittest into a separate object, Bugzilla 14801
    obj_start(mname);

    objmod->initfile(mname, NULL, m->toPrettyChars());
    genhelpers(m, true);
    objmod->termfile();

    obj_end(library, m->objfile);
```